marimo/_save/hash.py

```diff
@@ -154,20 +294,101 @@ def cache_attempt_from_hash(
     # Empty name, so we can match and fill in cell context on load.
     visitor = ScopedVisitor("", ignore_local=True)
     visitor.visit(module)
+    # Determine immediate references
+    refs = set(visitor.refs)
+
+    # Get stateful registers
+    # This is typically done in post execution hook, but it will not be called
+    # in script mode.
+    ctx = get_context()
+    ctx.ui_element_registry.register_scope(defs)
+    ctx.state_registry.register_scope(defs)
+
+    stateful_refs = normalize_and_extract_ref_state(
+        visitor.refs, refs, defs, cell_id, ctx
+    )
+
+    # usedforsecurity=False used to satisfy some static analysis tools.
+    hash_alg = hashlib.new(hash_type, usedforsecurity=False)
+    # Default type, means that there are no references at all.
+    cache_type: CacheType = "Pure"
 
     # Attempt content hash
-    valid_cache_sha = build_content_hash(graph, cell_id, visitor, defs)
-    if not valid_cache_sha:
+    if refs:
+        cache_type = "ContentAddressed"
+        hash_and_dequeue_content_refs(
+            hash_alg, cell_id, defs, refs, graph, pin_modules=pin_modules
+        )
+        # Determine _all_ additional relevant references
+        transitive_state_refs = graph.get_transitive_references(
+            visitor.refs, inclusive=False
+        )
+        refs |= set(
+            filter(
+                lambda ref: (
+                    ctx.state_registry.lookup(ref)
+                    or ctx.ui_element_registry.lookup(ref)
+                ),
+                transitive_state_refs,
+            )
+        )
+        # Need to run extract again for the expanded ref set.
+        stateful_refs |= normalize_and_extract_ref_state(
+            visitor.refs, refs, defs, cell_id, ctx
+        )
+
+        # Attempt content hash
+        hash_and_dequeue_content_refs(
+            hash_alg, cell_id, defs, refs, graph, pin_modules=pin_modules
+        )
```

Comment:
Can you leave a comment explaining this algorithm, and why it requires multiple passes?